plugin/action/throttle/redis_limiter.go

```diff
@@ -0,0 +1,171 @@
+package throttle
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ozontech/file.d/logger"
+	"github.com/ozontech/file.d/pipeline"
+)
+
+const (
+	keySuffix = "limit"
+)
+
+type redisLimiter struct {
+	mu    sync.Mutex
+	redis redisClient
+
+	// <keyPrefix>_
+	// will be used for bucket counter <keyPrefix>_<bucketID> and limit key <keyPrefix>_limit
+	keyPrefix bytes.Buffer
+
+	// contains values which will be used for incrementing remote bucket counter
+	// buckets will be flushed after every sync to contain only increment value
+	incrementLimiter *inMemoryLimiter
+
+	// contains global values synced from redis
+	totalLimiter *inMemoryLimiter
+}
+
+func NewRedisLimiter(
+	ctx context.Context,
+	redis redisClient,
+	throttleFieldName, throttleFieldValue string,
+	refreshInterval, bucketInterval time.Duration,
+	bucketCount int,
+	limit complexLimit,
+) *redisLimiter {
+	rl := &redisLimiter{
+		redis:            redis,
+		incrementLimiter: NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+		totalLimiter:     NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+	}
+
+	rl.keyPrefix = bytes.Buffer{}
+
+	// full name of keyPrefix will be $throttleFieldName_$throttleFieldValue_limit. `limit` added afterwards
+	rl.keyPrefix.WriteString(throttleFieldName)
+	rl.keyPrefix.WriteString("_")
+	rl.keyPrefix.WriteString(throttleFieldValue)
+	rl.keyPrefix.WriteString("_")
+
+	ticker := time.NewTicker(refreshInterval)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				rl.sync()
+			}
+		}
+	}()
+
+	return rl
+}
+
+func (l *redisLimiter) isAllowed(event *pipeline.Event, ts time.Time) bool {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// count increment from last sync
+	isAllowed := l.incrementLimiter.isAllowed(event, ts)
+	if isAllowed {
+		// count global limiter if distributed mode on and passed local limit
+		isAllowed = l.totalLimiter.isAllowed(event, ts)
+	}
+
+	return isAllowed
+}
+
+func (l *redisLimiter) sync() {
+	l.mu.Lock()
```

Comment:
It will lock `isAllowed()` function until sync is done. You may lock only for copy data in incrementLimiter and then do sync without locking

Reply:
After refactoring this mu redundant. I removed it.